File: NeetCode150/01. Arrays & Hashing/TwoSum.py
class Solution(object):
    def twoSum(self, nums, target):
        # Dictionary to store:
        # Key   -> number
        # Value -> index of that number
        seen = {}

        # Traverse the array with both index and value
        for i, num in enumerate(nums):

            # Find the number needed to reach the target
            complement = target - num

            # If the complement has already been seen,
            # we found the required pair
            if complement in seen:
                # Return the index of the complement
                # and the current index
                return [seen[complement], i]

            # Store the current number and its index
            # so it can be used for future complements
            seen[num] = i

        # If no valid pair exists, return an empty list
        return []

File: PatternWiseDSA/src/TwoPointers/TwoSum.java
package TwoPointers;

import java.util.HashMap;

//Leetcode link - https://leetcode.com/problems/two-sum/description/

public class TwoSum {
    public int[] twoSum(int[] nums, int target) {

        // Stores: value -> index
        HashMap<Integer, Integer> map = new HashMap<>();

        // Traverse array once
        for (int i = 0; i < nums.length; i++) {

            int current = nums[i];

            // Number needed to reach target
            int complement = target - current;

            // Check if complement already exists
            if (map.containsKey(complement)) {

                // Return indices
                return new int[]{map.get(complement), i};
            }

            // Store current number and its index
            map.put(current, i);
        }

        return new int[]{-1, -1};
    }
}
